src/oracle/middleware.js

```diff
@@ -0,0 +1,761 @@
+    body: {
+      "payment": 0,
+      "request_id": count,
+      "request_type": requestType
+    },
+    headers: {
+      "Content-Type": "application/json",
+      "X-Chainlink-EA-AccessKey": credentials["EI_IC_ACCESSKEY"],
+      "X-Chainlink-EA-Secret": credentials["EI_IC_SECRET"]
+    },
+    method: 'POST',
+  };
+
+  //try request with loop retries
+  for (let i = 0; i < SUBMIT_RETRIES; i++) {
+    try {
+      await axios.post(options.url, options.body, {
+        timeout: 60000,
+        proxy: false,
+        headers: options.headers,
+        httpAgent: new http.Agent({ keepAlive: false })
+      });
+      return
+    }
+    catch (err) {
+      console.error("JOB Request for " + jobId + " failed", err)
+    }
+  }
+}
+
+/**
+  * Function to get last 10 offers
+  * @param {*} follower offers and balances
+  * @returns a list of offers
+  */
+export const getOffers = async (follower) => {
+
+  let history = [];
+  let counter = 0;
+  let lastVisited = 0;
+
+  for await (const followerElement of iterateReverse(follower)) {
+
+    if (counter == 5){
+      break;
+    }
+
+    //if it is an offer status 
+    if (followerElement.value.updated == "offerStatus"){
+      //get id
+      let id = followerElement.value.status.id
+
+      //if a new and final state
+      if (id != lastVisited) {
+        //if it is not failed
+        if (!followerElement.value.status.hasOwnProperty("error")) {
+          history.push(followerElement.value);
+          counter++
+        }
+        lastVisited = id;
+      } 
+    }
+  }
+  return history
+}
+
+/**
+ * Function to get the latest submitted round
+ * @returns the latest round submitted
+ */
+export const getLatestSubmittedRound = async () => {
+  const unserializer = boardSlottingMarshaller(fromBoard.convertSlotToVal);
+  const leader = makeLeader(networkConfig.rpcAddrs[0]);
+
+  const follower = await makeFollower(
+    `:published.wallet.${FROM}`,
+    leader,
+    {
+      // @ts-expect-error xxx
+      unserializer,
+    },
+  );
+
+  //get offers
+  let offers = await getOffers(follower)
+
+  return Number(offers[0]["status"]["invitationSpec"]["invitationArgs"][0]["roundId"])
+}
+
+/**
+ * Function to check if submission was satisfied for a specific round
+ * @param {*} oracle oracle address
+ * @param {*} feedOfferId the offer id of the feed to check for
+ * @param {*} roundId the round Id which is checked
+ * @returns whether a submission was successful for a specific round
+ */
+export const checkSubmissionForRound = async (oracle, feedOfferId, roundId) => {
+  const unserializer = boardSlottingMarshaller(fromBoard.convertSlotToVal);
+  const leader = makeLeader(networkConfig.rpcAddrs[0]);
+
+  const follower = await makeFollower(
+    `:published.wallet.${oracle}`,
+    leader,
+    {
+      // @ts-expect-error xxx
+      unserializer,
+    },
+  );
+
+  //get offers
+  let offers = await getOffers(follower)
+  
+  //loop through offers starting from last offer
```

Comment:
So the logic here is, first getOffers and get the latest 5 offers. Then you use this list to check if a round was successful. 

Why don't you combine them instead of iterating twice?

Reply:
getOffers is used for multiple cases so it is better to leave it this way. Also the second loop would look 5 objects at most so not a big deal